### PR TITLE
Robust facultative cut generation callback

### LINKED ISSUE #177: Cut generation callback

### PR DESCRIPTION
**Pull request by guimarqu**
Coluna is now a real branch-**cut**-and-price !

Close #177 
Replace #256

### CODE REVIEW

File: src/Algorithm/conquer.jl [hunk 1]
```diff
@@ -154,18 +155,37 @@ function run!(algo::ColGenConquer, data::ReformData, input::ConquerInput)
...
-    if !to_be_pruned(node)
-        if algo.run_mastipheur 
-            @logmsg LogLevel(0) "Run IP restricted master heuristic."
-            TO.@timeit Coluna._to "RestMasterHeur" begin
-                heur_output = run!(
-                    algo.mastipheur, getmasterdata(data), OptimizationInput(nodestate)
-                )
-                update_all_ip_primal_solutions!(nodestate, getoptstate(heur_output))
-            end
+        update!(nodestate, getoptstate(colgen_output))
```
Comment (vitornesello) [inline src/Algorithm/conquer.jl:161]: Incorrect indentation here?

**Comment by guimarqu** [comment]
> Ok for me for now.
> 
> In the future, we need a separate algorithm per cut callback because priorities may be different.
> 
> Also for the moment, separation of only facultative cuts is possible. Core (essential) cut separation should be integrated to every algorithm which can change the primal IP solution (column generation, primal heuristics, etc.).

Ok we will have to check what JuMP callbacks allow us to do.

**Comment by rrsadykov** [comment]
@guimarqu Are you sure, CVRP test is working. I have this on my computer:

```
Coluna
Version 0.3 - https://github.com/atoptima/Coluna.jl
***************************************************************************************
**** BaB tree root node
**** Local DB = -Inf, global bounds : [ -Inf , 784.0000 ], time = 0.00 sec.
***************************************************************************************
  <it=  1> <et= 0.10> <mst= 0.01> <sp= 0.09> <cols= 1> <DB=-1549280.0000> <mlp=20000.0000> <PB=784.0000>
  <it=  2> <et= 0.23> <mst= 0.00> <sp= 0.13> <cols= 1> <DB=-1222693.3333> <mlp=20000.0000> <PB=784.0000>
  <it=  3> <et= 0.49> <mst= 0.00> <sp= 0.26> <cols= 1> <DB=-1190892.3595> <mlp=20000.0000> <PB=784.0000>
  <it=  4> <et= 0.58> <mst= 0.00> <sp= 0.09> <cols= 1> <DB=-702483.7702> <mlp=20000.0000> <PB=784.0000>
  <it=  5> <et= 0.65> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=-451340.8924> <mlp=20000.0000> <PB=784.0000>
  <it=  6> <et= 0.78> <mst= 0.00> <sp= 0.12> <cols= 1> <DB=-451340.8924> <mlp=20000.0000> <PB=784.0000>
  <it=  7> <et= 0.84> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=-405612.7059> <mlp=20000.0000> <PB=784.0000>
  <it=  8> <et= 0.90> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-405612.7059> <mlp=20000.0000> <PB=784.0000>
  <it=  9> <et= 0.95> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-252858.3529> <mlp=20000.0000> <PB=784.0000>
  <it= 10> <et= 1.00> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-252858.3529> <mlp=20000.0000> <PB=784.0000>
  <it= 11> <et= 1.05> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-213874.5882> <mlp=20000.0000> <PB=784.0000>
  <it= 12> <et= 1.11> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-213874.5882> <mlp=20000.0000> <PB=784.0000>
  <it= 13> <et= 1.17> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-213874.5882> <mlp=20000.0000> <PB=784.0000>
  <it= 14> <et= 1.22> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-213874.5882> <mlp=20000.0000> <PB=784.0000>
  <it= 15> <et= 1.26> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-213874.5882> <mlp=20000.0000> <PB=784.0000>
  <it= 16> <et= 1.31> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-213874.5882> <mlp=20000.0000> <PB=784.0000>
  <it= 17> <et= 1.36> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-179412.9412> <mlp=20000.0000> <PB=784.0000>
  <it= 18> <et= 1.40> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-179412.9412> <mlp=20000.0000> <PB=784.0000>
  <it= 19> <et= 1.44> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-179412.9412> <mlp=20000.0000> <PB=784.0000>
  <it= 20> <et= 1.49> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-176837.1765> <mlp=20000.0000> <PB=784.0000>
  <it= 21> <et= 1.56> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-176837.1765> <mlp=20000.0000> <PB=784.0000>
  <it= 22> <et= 1.62> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-176837.1765> <mlp=20000.0000> <PB=784.0000>
  <it= 23> <et= 1.69> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=-176837.1765> <mlp=20000.0000> <PB=784.0000>
  <it= 24> <et= 1.76> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=-176837.1765> <mlp=20000.0000> <PB=784.0000>
  <it= 25> <et= 1.83> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=-176837.1765> <mlp=20000.0000> <PB=784.0000>
  <it= 26> <et= 1.90> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=-176837.1765> <mlp=20000.0000> <PB=784.0000>
  <it= 27> <et= 1.95> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-174031.4482> <mlp=20000.0000> <PB=784.0000>
  <it= 28> <et= 2.01> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-174031.4482> <mlp=20000.0000> <PB=784.0000>
  <it= 29> <et= 2.07> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-174031.4482> <mlp=20000.0000> <PB=784.0000>
  <it= 30> <et= 2.14> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-174031.4482> <mlp=20000.0000> <PB=784.0000>
  <it= 31> <et= 2.21> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=-174031.4482> <mlp=20000.0000> <PB=784.0000>
  <it= 32> <et= 2.27> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-174031.4482> <mlp=20000.0000> <PB=784.0000>
  <it= 33> <et= 2.32> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-174031.4482> <mlp=20000.0000> <PB=784.0000>
  <it= 34> <et= 2.38> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-174031.4482> <mlp=20000.0000> <PB=784.0000>
  <it= 35> <et= 2.43> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-174031.4482> <mlp=19335.7839> <PB=784.0000>
  <it= 36> <et= 2.48> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-174031.4482> <mlp=18380.6361> <PB=784.0000>
  <it= 37> <et= 2.52> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-174031.4482> <mlp=16431.1887> <PB=784.0000>
  <it= 38> <et= 2.57> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-145082.4566> <mlp=13528.4127> <PB=784.0000>
  <it= 39> <et= 2.61> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-145082.4566> <mlp=12536.4959> <PB=784.0000>
  <it= 40> <et= 2.67> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-145082.4566> <mlp=12364.9735> <PB=784.0000>
  <it= 41> <et= 2.71> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-145082.4566> <mlp=12306.2892> <PB=784.0000>
  <it= 42> <et= 2.76> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-145082.4566> <mlp=12306.2892> <PB=784.0000>
  <it= 43> <et= 2.80> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-145082.4566> <mlp=12207.8821> <PB=784.0000>
  <it= 44> <et= 2.84> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-145082.4566> <mlp=12056.6844> <PB=784.0000>
  <it= 45> <et= 2.89> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-145082.4566> <mlp= 8176.9026> <PB=784.0000>
  <it= 46> <et= 2.93> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-145082.4566> <mlp= 8100.3728> <PB=784.0000>
  <it= 47> <et= 2.97> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-145082.4566> <mlp= 7963.4118> <PB=784.0000>
  <it= 48> <et= 3.02> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-145082.4566> <mlp= 6909.7299> <PB=784.0000>
  <it= 49> <et= 3.06> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-143379.7154> <mlp= 5774.0183> <PB=784.0000>
  <it= 50> <et= 3.11> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-143379.7154> <mlp= 2047.2166> <PB=784.0000>
  <it= 51> <et= 3.16> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-7757.2499> <mlp= 2047.2166> <PB=784.0000>
  <it= 52> <et= 3.22> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-4586.4815> <mlp= 2043.1028> <PB=784.0000>
  <it= 53> <et= 3.28> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-2325.1861> <mlp= 2037.0814> <PB=784.0000>
  <it= 54> <et= 3.32> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=-2325.1861> <mlp= 2035.7886> <PB=784.0000>
  <it= 55> <et= 3.38> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=-2325.1861> <mlp= 2028.9300> <PB=784.0000>
  <it= 56> <et= 3.42> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-2325.1861> <mlp= 2026.7256> <PB=784.0000>
  <it= 57> <et= 3.47> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=-2118.5824> <mlp= 2016.1670> <PB=784.0000>
  <it= 58> <et= 3.55> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=-1606.5263> <mlp= 2009.6413> <PB=784.0000>
  <it= 59> <et= 3.61> <mst= 0.00> <sp= 0.07> <cols= 1> <DB= -489.7613> <mlp= 2009.6413> <PB=784.0000>
  <it= 60> <et= 3.68> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=  203.0860> <mlp= 2009.6413> <PB=784.0000>
  <it= 61> <et= 3.75> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=  203.0860> <mlp= 2007.2693> <PB=784.0000>
  <it= 62> <et= 3.80> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=  203.0860> <mlp= 1999.9451> <PB=784.0000>
  <it= 63> <et= 3.87> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=  203.0860> <mlp= 1997.3599> <PB=784.0000>
  <it= 64> <et= 3.91> <mst= 0.00> <sp= 0.04> <cols= 1> <DB=  203.0860> <mlp= 1991.6749> <PB=784.0000>
  <it= 65> <et= 3.98> <mst= 0.00> <sp= 0.07> <cols= 1> <DB=  203.0860> <mlp= 1990.5424> <PB=784.0000>
  <it= 66> <et= 4.03> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=  203.0860> <mlp= 1988.9327> <PB=784.0000>
  <it= 67> <et= 4.08> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=  203.0860> <mlp= 1981.6056> <PB=784.0000>
  <it= 68> <et= 4.13> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=  203.0860> <mlp= 1976.7482> <PB=784.0000>
  <it= 69> <et= 4.18> <mst= 0.00> <sp= 0.05> <cols= 1> <DB=  390.8991> <mlp= 1972.3437> <PB=784.0000>
  <it= 70> <et= 4.25> <mst= 0.00> <sp= 0.06> <cols= 1> <DB=  393.0052> <mlp= 1969.2213> <PB=784.0000>
  <it= 71> <et= 4.30> <mst= 0.00> <sp= 0.05> <cols= 1> <DB= 1233.4001> <mlp= 1968.1947> <PB=784.0000>
[ Info: Dual bound reached primal bound.
Node is already conquered. No children will be generated
 ────────────────────────────────────────────────────────────────────────────
                                     Time                   Allocations      
                             ──────────────────────   ───────────────────────
      Tot / % measured:           30.9s / 13.9%           4.42GiB / 96.8%    

 Section             ncalls     time   %tot     avg     alloc   %tot      avg
 ────────────────────────────────────────────────────────────────────────────
 Coluna                   1    4.30s   100%   4.30s   4.28GiB  100%   4.28GiB
   SolveLpForm           71   44.9ms  1.04%   632μs   4.68MiB  0.11%  67.5KiB
   Cleanup columns       71    438μs  0.01%  6.17μs   33.5KiB  0.00%     483B
   Restore states         1   28.6μs  0.00%  28.6μs      272B  0.00%     272B
   Store states           1   14.1μs  0.00%  14.1μs      528B  0.00%     528B
 ────────────────────────────────────────────────────────────────────────────
[ Info: Terminated
[ Info: Primal bound: 784.0
[ Info: Dual bound: 784.0
Test Summary: |
play cvrp     | No tests

```

**Comment by guimarqu** [comment]
I just noticed that I have two repo ColunaDemos and ColunaDemos.jl in my dev folder. The demo is not up to date.

**Comment by rrsadykov** [comment]
@guimarqu  I have updated ColunaDemos, but still it does not work for me. Also, in `cvrp_tests()` you do not test anything.

**Comment by guimarqu** [comment]
It works now

**Pull request merged**